### PR TITLE
Workflow position email notifications

### PR DESCRIPTION
**Pull request by miles-smith**
This PR somewhat indirectly addressess the request for different mail content/terminology to be used when an application manager receives notification about a project being closed (Plan.IO 27280).

I've approached this from a slightly more abstracted perspective and the bulk of the PR is a lot of groundwork and foundations for sending emails to interested parties when a project moves to *any* state (and in a way that helps to decouple emails from `Notification`s and remove side effect logic from the persistence layer).

The use of localisation over branching logic feels like a nice touch, and could help in the long run when things like views diverge to the point where having a single template feels constraining and hurts readability (from e.g. excessive branching), but I'm slightly concious about it...
- Is it correct and consistent in terms of favoured approach to these problems within the team and our other codebases? (although I know we've done so on at least one other project, to great effect).
- Accessibility to the next dev; can they just pick it up and run, or will they get bogged down in layers of indirection/abstraction?
- Am I using a scalpel to spread butter?

So, thoughts very much welcome on this one!

Commits:
- Add `project_state_name` delegate.
- Add framework for sending project state change emails
- Replicate mail generated from rejected Notification
- Boyscouting - clearer variable names
- Backport translations fix for state names
- Use localisation for varying email content for different target audiences
- Add application/eoi specific rejected email
- Add an ODR specific version of the project rejected email
- Wire up post transition email notifications
- Project rejected `Notification`s should not generate email
- Do not email about state changes under certain condiftions

### CODE REVIEW

File: test/mailers/concerns/workflow/mail_test.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    end
+
+    test 'default transition email, with comment' do
+      project = projects(:dummy_project)
+
+      email = ProjectsMailer.with(
+        project:      project,
+        user:         project.owner,
+        current_user: project.owner,
+        comments:     'RAWR!'
```
Comment (kenny-lee-1) [inline test/mailers/concerns/workflow/mail_test.rb:37]: 🦁

File: app/controllers/concerns/workflow/controller.rb [hunk 2]
```diff
@@ -86,6 +84,19 @@ def send_transition_email
       end
     end
 
+    # Who should receive emails about changes to a `project`s workflow position.
+    # In lieu of a subscription based model where users can manage their own desired status updates.
+    # NOTE: yuk.
+    def notifiable_users
```
Comment (kenny-lee-1) [inline app/controllers/concerns/workflow/controller.rb:90]: So I guess ODR user only needs an email to inform them they need to action something and do not care about status updates. And then we cycle back to application managers being pseudo users.

File: app/controllers/concerns/workflow/controller.rb [hunk 3]
```diff
@@ -86,6 +84,19 @@ def send_transition_email
...
+      # Does the user performing the transition need notifying?
+      scope = @project.users.where.not(id: current_user)
+
+      # Copied from original project rejected notification...
+      scope = scope.where.not(id: @project.users.odr_users) if @project.application?
```
Comment (kenny-lee-1) [inline app/controllers/concerns/workflow/controller.rb:95]: would this be notifying ODR for EOI, MBIS, CAS & other future application types?

File: test/mailers/concerns/workflow/mail_test.rb [hunk 4]
```diff
@@ -99,5 +99,36 @@ class MailTest < ActionMailer::TestCase
...
+
+      project.assigned_user = user
+      project.save!(validate: false)
+
+      project.project_states.build(
+        state: workflow_states(:rejected),
+        user: user
+      ).save!(validate: false)
+
+      email = ProjectsMailer.with(
```
Comment (kenny-lee-1) [inline test/mailers/concerns/workflow/mail_test.rb:115]: should the building of the rejected state automatically trigger ProjectMailer?

**Comment by kenny-lee-1** [comment]
I think this is good approach and as you say there's precedent elsewhere. There's always going to be level of developer onboarding when looking at unfamiliar code. I've made a few comments.

File: app/controllers/concerns/workflow/controller.rb (hunk 3 above)
Comment (miles-smith) [reply app/controllers/concerns/workflow/controller.rb:95]: "ODR" as in the role, or as in the collective team of people?

Hypothetically speaking, if a given user has some kind of `Grant` on the `project`, and there is an appropriate method on the mailer that would handle the required mail out, then they would be mailed. 

This is one of the reasons why the generic mail method (`transitioned`) is *not* currently targetable via the `state_changed` mail method; we're not looking to spam users with every state change. 

Under the current configuration and available mail methods for state changes, the only mailing that will happen via this route is for projects of any kind being rejected, which is no different from what is currently happening via `Project#status_change_notifier` and `Project#odr_rejected_notification` as far as I can tell.

The exception for ODR users + applications, was bourne out of one user not wanting to receive an email at a specific point (a very specific edge case, which is simply being reimplemented here). We can add/tweak as needed but should be mindful that this is a smell that will probably grow over time.

This really feeds into the idea of having users manage the types of emails/notifications they are interested in receiving being a potentially beneficial bit of development; it empowers users to tune into what they want, and saves us having to bloat the codebase with conditional logic to cover edge case after edge case.

File: test/mailers/concerns/workflow/mail_test.rb (hunk 4 above)
Comment (miles-smith) [reply test/mailers/concerns/workflow/mail_test.rb:115]: Depends on who you ask.

I'm not a fan of side effect logic in the persistence layer. It has no bearing on what is required for saving a record, it can encourage tight coupling between components, and in some cases can lead to models having more LoC around mailing and notifications than any other actual behaviour it exists for. Generally, I'd rather interested parties to act in response to an event such as a record saving (e.g. in a pub/sub paradigm), or whatever is coordinating business logic (be it a controller, job, service object, command, Rake task etc.) to orchestrate things as required.

YMMV.

File: app/controllers/concerns/workflow/controller.rb (hunk 2 above)
Comment (miles-smith) [reply app/controllers/concerns/workflow/controller.rb:90]: Sorry, I'm not sure I follow you on this one? Care to elaborate?

File: test/mailers/concerns/workflow/mail_test.rb (hunk 1 above)
Comment (miles-smith) [reply test/mailers/concerns/workflow/mail_test.rb:37]: 😹

File: app/controllers/concerns/workflow/controller.rb (hunk 3 above)
Comment (miles-smith) [reply app/controllers/concerns/workflow/controller.rb:95]: Yep, probably.

I'm sure they're similarly not interested in the cases highlighted in your initial comment, but the ticket that asked for an exception didn't specify those, so they got exactly what was asked for 😉

File: app/controllers/concerns/workflow/controller.rb (hunk 2 above)
Comment (kenny-lee-1) [reply app/controllers/concerns/workflow/controller.rb:90]: Sorry. I think my point is on the other comment around our ODR user group size of one

File: test/mailers/concerns/workflow/mail_test.rb (hunk 4 above)
Comment (kenny-lee-1) [reply test/mailers/concerns/workflow/mail_test.rb:115]: Yep I wasn't quite sure if we were expecting the state change to trigger the mail for us.

**Review by kenny-lee-1 — APPROVED** [review]

File: app/controllers/concerns/workflow/controller.rb (hunk 2 above)
Comment (miles-smith) [reply app/controllers/concerns/workflow/controller.rb:90]: 👍
